lib/Sema/Constraint.cpp

```diff
@@ -337,22 +338,40 @@ void Constraint::print(llvm::raw_ostream &Out, SourceManager *sm) const {
     if (isIsolated())
       Out << " (isolated)";
 
+    firstLocator = Locator;
     if (Locator) {
       Out << " [[";
       Locator->dump(sm, Out);
       Out << "]]";
     }
     Out << ":\n";
-
-    interleave(getNestedConstraints(),
+    
+    // Sort constraints by favored, unmarked, disabled
+    // for printing only.
+    std::vector<Constraint *> sortedConstraints(getNestedConstraints().begin(),
+                                                getNestedConstraints().end());
+    llvm::sort(sortedConstraints,
+               [](const Constraint *lhs, const Constraint *rhs) {
+                 if (lhs->isFavored() != rhs->isFavored())
+                     return lhs->isFavored();
+                 if (lhs->isDisabled() != rhs->isDisabled())
+                     return rhs->isDisabled();
+                 return false;
+               });
+
+    interleave(sortedConstraints,
                [&](Constraint *constraint) {
                  if (constraint->isDisabled())
                    Out << ">  [disabled] ";
                  else if (constraint->isFavored())
                    Out << ">  [favored]  ";
                  else
                    Out << ">             ";
+                 if (constraint->getLocator() == firstLocator) {
+                   constraint->print(Out, sm, true);
```

Comment:
Nit: this could be simplified into `constraint->print(Out, sm, /*skipLocator=*/constraint->getLocator() == firstLocator);`

Note that every time we have to use constants such as `true|false` or numbers the convention is to add a comment with the name of the parameter, so it's easier to understand the meaning of the literal while reading the call itself.